docs/rfc-1-staking-contract.adoc

```diff
@@ -0,0 +1,381 @@
+require any owner’s keys at all. The stake delegation should also optimize the
+network throughput without compromising the security of the owners’ stake.
+
+This proposal aims at implementing a minimum viable staking contract version
+allowing to support legacy KEEP and NU stakes, as well as native T delegations.
+The functionality of the staking contract can be further extended by the
+upgradeability of the contract code.
+
+=== Terminology
+
+Owner:: An address owning T tokens and tokens that could be converted to
+T: KEEP, NU, including grants. An owner is the ultimate holder of the tokens.
+Before stake delegation, the owner has full control over the tokens, and the
+tokens are returned to the owner after stake delegation has finished.
+Owner’s participation is not required in the day-to-day operations on the
+stake, so cold storage can be accommodated to the maximum extent.
+
+Operator:: An address of a party authorized to operate in the network on behalf
+of a given owner. The operator handles the everyday operations on the delegated
+stake without actually owning the staked tokens. An operator can not simply
+transfer away delegated tokens, however, it should be noted that the operator’s
+misbehavior may result in slashing tokens and thus the entire staked amount is
+indeed at stake.
+
+Beneficiary:: An address where the rewards for participation are sent, earned by
+the operator, on behalf of the owner. A beneficiary doesn’t sign or publish any
+protocol-relevant transactions, but any currency or tokens earned by the
+operator will be transferred to the beneficiary.
+
+Authorizer:: An address appointed by an owner to authorize applications on
+behalf of them. An application must be approved by the authorizer before the
+operator is eligible to participate.
+
+Delegated stake:: An owner’s staked tokens, delegated to the operator by the
+owner. Delegation enables KEEP owners to have their wallets offline and their
+stake operated by operators on their behalf.
+
+== Specification
+
+=== Functionality
+
+==== Delegating a stake
+
+Tokens are delegated by the owner. During the delegation, the owner needs to
+appoint an operator, beneficiary, and authorizer. The owner may decide to
+delegate just a portion of their tokens. The owner may delegate multiple times
+to different operators. For simplicity, the operator address can not be reused
+between delegations even if the delegation has been recovered. The only
+exception is when the delegation has been canceled. To protect against griefing
+by frontrunning stake delegation transactions and reserving operator addresses,
+staking contract should require a governable minimum stake amount for each
+delegation.
+
+Staking contract binds the owner with the operator, beneficiary, and authorizer
+for the given stake delegation and records the stake delegation amount.
+Delegation is set to active after 2 hours, once the cancellation window passes.
```

Comment:
If we have a separate undelegation period we should have the 2 hour cancellation window so that mistakenly delegating to the wrong operator address doesn't lock people's capital up for a long time.

Reply:
I mean in current PR there is no lock right after stake, lock applies only after second tx with allocation to app

Reply:
Like two layers: one is deposit/withdrawing tokens (no lock itself). Second layer - allocation/authorization per app. In that case max(any allocation) will be locked. We can consider 2 hours for cancelling allocation but if reason only to save from accident with initial deposit then owner will be ably to withdraw without waiting

Reply:
Allocations are better controlled (has to be approved by governance etc.) than operator addresses (any arbitrary address works) so I don't think we need a cancellation for allocations.

Reply:
I think @vzotova suggests removing the cancelation timeout and let the stake be withdrawn as soon as all authorized applications release it or, if there were no authorizations yet, let to undelegate the stake immediately. If so it depends on the result of [this](https://github.com/threshold-network/solidity-contracts/pull/15/files/c1e8a0ba7479ffb23fd6881d4bc08f113a8a2c2c#r703412443) discussion.